# 🧐 Evaluate Modell
### Insurance Daten laden

In [1]:
# IMPORT LIBRARIES
import pandas as pd
from sklearn.model_selection import train_test_split

# LOAD DATA
df = pd.read_csv("03_insurance_data.csv")

### 🐼 One-Hot-Encoding mit Pandas

In [2]:
df_encoded = pd.get_dummies(df, columns=['sex', 'smoker', 'region'], dtype=int)

In [3]:
# FEATURES & TARGET
X = df_encoded.drop(['charges'], axis=1)
Y = df_encoded['charges']

### ✂️ Insurance Daten aufteilen
Daten splitten in 70% Training und 30% Testdaten.

In [4]:
# DATEN IN TRAIN UND TEST SPLITTEN
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=1234)

### 🏋🏻‍♂️ Modell trainieren
Das ML-Modell mit einem Algorithmus deiner Wahl trainieren, um Vorhersagen treffen zu können.

In [5]:
# IMPORT LIBRARIES
from sklearn.linear_model import LinearRegression

# MODELL ERSTELLEN
lr = LinearRegression()

# MODELL TRAINIEREN
lr.fit(X_train, Y_train)

# PREDICTIONS MACHEN
Y_pred_lr = lr.predict(X_test)

In [6]:
# IMPORT LIBRARIES
from sklearn.neighbors import KNeighborsRegressor

# MODELL ERSTELLEN
nr = KNeighborsRegressor()

# MODELL TRAINIEREN
nr.fit(X_train, Y_train)

# PREDICTIONS MACHEN
Y_pred_nr = nr.predict(X_test)

### 🧮 Kennzahlen berechnen

In [7]:
# IMPORT LIBRARIES
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

# KENNZAHLEN FÜR LR BERECHNEN
mse_lr = mean_squared_error(Y_test, Y_pred_lr)
rmse_lr = np.sqrt(mse_lr)
mae_lr = mean_absolute_error(Y_test, Y_pred_lr)
r2_lr = r2_score (Y_test, Y_pred_lr)

# KENNZAHLEN FÜR NR BERECHNEN
mse_nr = mean_squared_error(Y_test, Y_pred_nr)
rmse_nr = np.sqrt(mse_nr)
mae_nr = mean_absolute_error(Y_test, Y_pred_nr)
r2_nr = r2_score (Y_test, Y_pred_nr)

# OUTPUT RESULTS
print(f"MSE LR: {mse_lr:.2f}")
print(f"MSE NR: {mse_nr:.2f}")

print(f"RMSE LR: {rmse_lr:.2f}")
print(f"RMSE NR: {rmse_nr:.2f}")

print(f"MAE LR: {mae_lr:.2f}")
print(f"MAE NR: {mae_nr:.2f}")

print(f"R² LR: {r2_lr:.2f}")
print(f"R² NR: {r2_nr:.2f}")

MSE LR: 37250298.44
MSE NR: 109378732.56
RMSE LR: 6103.30
RMSE NR: 10458.43
MAE LR: 4191.18
MAE NR: 7067.37
R² LR: 0.74
R² NR: 0.23


- MSE: Niedriger ist besser. Das LR-Modell hat einen deutlich niedrigeren Wert
- RMSE: Niedriger ist besser. Das LR-Modell ist hier ebenfalls besser.
- MAE: Niedriger ist besser. Das LR-Modell ist auch hier besser.
- R²: Hier ist höher besser. Das LR-Modell liegt hier bei 74%, NR-Modell nur bei 23%
## Daher ist das LR-Modell für den vorliegenden Datensatz besser geeignet.